wiki/Concept-Starting-and-Orchestrating-Bots.md

```diff
@@ -1,89 +0,0 @@
-
-### Higher Abstraction Frameworks
-
-There are some frameworks using one (or more) of these basic approaches but are one abstraction layer higher for the user. In particular, pusher and socket.io were looked at.
-Pusher is based on a publish subscribe approach simular to SSE and it seems to be very easy to use. The biggest disadvantage was that it is not open-source but ultimatly one would have to pay for it. Socket.io is a highly popular (> 4 Mio. weekly npm downloads) open source framework. It is mainly based on websocket communication but is also able to seamlessly switch to a long polling approach in the background if websockets do not work.
-
-### Decision
-
-In the end, we decided for socket.io as it is open-cource, well-supported, 'easy' to use, robust and websocket based enabling a bidirectional communication.
-To get started it is recommended reading [this](https://socket.io/docs/v4/index.html) introduction to socket.io. Especially these two subpages [1](https://socket.io/docs/v4/server-socket-instance/) & [2](https://socket.io/docs/v4/client-socket-instance/) are relevant for this usecase.
-
-## How the new approach with socket.io was implemented in the ark automate product
-
-**Desktop app:**
-Here we implemented a CLI that reads the userId the user enters, saves the userId and uses the userId as an authentication for the communication with the web app server. Once started, the desktop app connects with the server by using a socket connection. A socket connection is this bidirectional connection that exist between every client and server. Moreover we use the userId the user entered and ask the server if this socket connection can join the room userId (we have one room for every userId). Once the socket connection was added to the userId room we wait for robots to be executed. Because we are in the userId room we receive robot execution jobs of web frontends that are connected to the same room.
-
-**Database/Server:**
-We implemented a jobs collection in MongoDB as well as a Mongoose jobs model. Every job has a robotId, a userId, a status (waiting/executing/success/failed) and an array of parameters that contains the arguments the user entered in the web frontend when starting the robot execution.
-
-**Server:**
-Sets up a server and socket instance, establishes socket connection with the web frontend and the desktop app, groups sockets by userIds (by using the room concept), reacts on robot execution commands and forwards this command to the desktop app and updates the jobs collection in MongoDB continiously.
-
-**Web Frontend:**
-Connects with the server using a socket connection. Also, like the dektop app, we join a userId specific room whenever the robot overview is rendered. Additionally, we send a robot execution job to the backend when the user clicks on the play button in the robot container.
```

Comment:
I think a part regarding how the socket.io implementation works would have been quite nice. The local client readme does not explain it with as much detail like here and it would for sure help users understand the implementation better. But seeing the time this is out of scope.

Reply:
reimplemented!